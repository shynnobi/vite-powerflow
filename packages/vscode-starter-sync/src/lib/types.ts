export type Status = 'sync' | 'warning' | 'error';

export interface CheckResult {
  status: Status;
  message: string;
  commitCount: number;
}
