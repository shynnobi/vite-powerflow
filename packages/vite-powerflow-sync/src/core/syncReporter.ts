import * as path from 'path';

import { formatBaseline } from './baselineFormatter.js';
import { readPackageInfo } from './packageReader.js';
import { CheckResult, PackageLabel, SyncCheckConfig } from './types.js';

export function formatPackageStatus(label: PackageLabel, result: CheckResult): string {
  const versionInfo = result.packageVersion ? ` (v${result.packageVersion})` : '';
  let lines: string[] = [];

  lines.push(`📦 [${label}]${versionInfo}`);

  if (result.status === 'error') {
    lines.push(`   ❌ Check failed - ${result.message}`);
    return lines.join('\n');
  }

  if (result.commitCount === 0) {
    const commitInfo = result.baselineCommit
      ? ` - baseline ${result.baselineCommit.substring(0, 7)}`
      : '';
    lines.push(`   ✅ Package in sync${commitInfo}`);
    return lines.join('\n');
  }

  if (result.changeset) {
    lines.push(`   📄 Changeset: ${result.changeset.fileName} (${result.changeset.bumpType})`);

    if (result.commits && result.commits.length > 0) {
      const coveredCommits = result.coveredCommits ?? [];
      const notCoveredCommits = result.notCoveredCommits ?? [];
      const totalCommits = result.commits.length;

      lines.push(`   📊 Coverage: ${coveredCommits.length}/${totalCommits} commits covered`);

      if (notCoveredCommits.length === 0) {
        lines.push(`   🎯 Ready for release`);
      } else {
        const plural = notCoveredCommits.length > 1 ? 'commits' : 'commit';
        lines.push(`   ⚠️  ${notCoveredCommits.length} ${plural} NOT covered by changeset:`);
        notCoveredCommits.forEach(commit => {
          lines.push(`      • ${commit.sha.substring(0, 7)} ${commit.message}`);
        });
        lines.push(`   ⚠️ Require changeset update`);
      }
    } else {
      lines.push(`   🎯 Ready for release`);
    }
  } else {
    lines.push(`   ❌ No changeset found`);
    const plural = result.commitCount > 1 ? 'commits' : 'commit';
    lines.push(`   📋 ${result.commitCount} unreleased ${plural} need changeset:`);

    if (result.commits) {
      const maxShow = 5;
      const commitsToShow = result.commits.slice(0, maxShow);
      commitsToShow.forEach(commit => {
        lines.push(`      • ${commit.sha.substring(0, 7)} ${commit.message}`);
      });

      if (result.commits.length > maxShow) {
        lines.push(`      • ... and ${result.commits.length - maxShow} more`);
      }
    }
    lines.push(`   ⚠️ Require changeset (${result.commitCount} unreleased ${plural})`);
  }

  return lines.join('\n');
}

export function formatGlobalStatus(
  results: { label: PackageLabel; result: CheckResult }[]
): string[] {
  if (results.length === 0) return [];

  let syncCount = 0;
  let pendingCount = 0;
  let warningCount = 0;
  let errorCount = 0;

  for (const { result } of results) {
    if (result.status === 'error') {
      errorCount++;
    } else if (result.commitCount === 0) {
      syncCount++;
    } else if (result.changeset) {
      const notCoveredCommits = result.notCoveredCommits ?? [];
      if (notCoveredCommits.length > 0) {
        warningCount++;
      } else {
        pendingCount++;
      }
    } else {
      warningCount++;
    }
  }

  const totalPackages = results.length;
  const packageWord = totalPackages > 1 ? 'packages' : 'package';

  let globalStatus = '';
  let globalEmoji = '';

  if (errorCount > 0) {
    globalStatus = 'ERROR';
    globalEmoji = '🔴';
  } else if (warningCount > 0) {
    globalStatus = 'WARNING';
    globalEmoji = '🟠';
  } else if (pendingCount > 0) {
    globalStatus = 'PENDING';
    globalEmoji = '🟡';
  } else {
    globalStatus = 'SYNC';
    globalEmoji = '🟢';
  }

  const result = [`🔄 Status: ${globalEmoji} ${globalStatus}`];

  let summary = '';
  if (errorCount + warningCount === 0) {
    summary = `📋 Summary: All ${totalPackages} ${packageWord} ready`;
  } else if (syncCount + pendingCount === 0) {
    summary = `📋 Summary: All ${totalPackages} ${packageWord} need attention`;
  } else {
    const readyCount = syncCount + pendingCount;
    const needAttentionCount = errorCount + warningCount;
    const readyWord = readyCount > 1 ? 'packages' : 'package';
    const needAttentionWord = needAttentionCount > 1 ? 'packages' : 'package';
    summary = `📋 Summary: ${readyCount} ${readyWord} ready, ${needAttentionCount} ${needAttentionWord} need attention`;
  }
  result.push(summary);

  const changesetMap = new Map<string, { label: PackageLabel; bumpType: string }[]>();

  for (const { label, result: pkgResult } of results) {
    if (pkgResult.changeset) {
      const fileName = pkgResult.changeset.fileName;
      if (!changesetMap.has(fileName)) {
        changesetMap.set(fileName, []);
      }
      changesetMap.get(fileName)!.push({ label, bumpType: pkgResult.changeset.bumpType });
    }
  }

  const multiPackageChangesets = Array.from(changesetMap.entries()).filter(
    ([, packages]) => packages.length > 1
  );

  if (multiPackageChangesets.length > 0) {
    for (const [fileName, packages] of multiPackageChangesets) {
      const packageList = packages
        .map(({ label, bumpType }) => `${label} (${bumpType})`)
        .join(', ');
      result.push(`🔗 ${fileName}: ${packageList}`);
    }
  }

  return result;
}

export function formatSyncOutput(
  results: { label: PackageLabel; result: CheckResult }[]
): string[] {
  const lines: string[] = [];

  const now = new Date();
  const pad = (n: number) => n.toString().padStart(2, '0');
  // Use local time for user-facing output; tests normalize this line for snapshots
  const timestamp = `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())} ${pad(
    now.getHours()
  )}:${pad(now.getMinutes())}:${pad(now.getSeconds())}`;

  lines.push(`🔄 Sync Status Report - [${timestamp}]`);
  lines.push('═══════════════════════════════════════════════════════════');
  lines.push('');

  for (const { label, result } of results) {
    const statusLines = formatPackageStatus(label, result).split('\n');
    lines.push(...statusLines);
    lines.push('');
  }

  lines.push('═══════════════════════════════════════════════════════════');
  const summaryLines = formatGlobalStatus(results);
  if (summaryLines.length > 0) {
    lines.push(...summaryLines);
  }

  lines.push('');
  lines.push('');

  return lines;
}

export async function formatBaselineLog(
  config: SyncCheckConfig,
  baseline: string,
  workspaceRoot: string
): Promise<string> {
  const shortBaseline = formatBaseline(baseline);
  let message = `📦 [${config.label}] Checking against baseline (commit/tag ${shortBaseline})`;

  if (config.label === PackageLabel.Starter) {
    try {
      const templatePackagePath = path.join(workspaceRoot, 'packages/cli/template/package.json');
      const templatePkg = await readPackageInfo(templatePackagePath);
      if (templatePkg?.version) {
        message = `📦 [Starter] Checking against CLI template baseline (commit ${shortBaseline}, version ${templatePkg.version})`;
      }
    } catch {}
  }

  return message;
}
