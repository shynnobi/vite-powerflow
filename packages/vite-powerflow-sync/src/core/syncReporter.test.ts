import { describe, expect, test, vi } from 'vitest';

import { formatBaselineLog, formatSyncOutput } from './syncReporter';
import { CheckResult, SyncCheckConfig } from './types';

// Mock the packages module
vi.mock('./packageReader.js', () => ({
  readPackageInfo: (path: string) => {
    if (path.includes('template/package.json')) {
      return Promise.resolve({ version: '1.0.0' });
    }
    return Promise.resolve(null);
  },
}));

// Utility to normalize time header in reports for stable snapshots
const normalizeReport = (lines: string[]): string[] => {
  const out = lines.slice();
  if (out.length > 0) {
    out[0] = out[0].replace(/\[[^\]]*\]/, '[TIME]');
  }
  return out;
};

// No fake timers needed; tests normalize timestamp in snapshots

describe('formatSyncOutput', () => {
  test('should show all packages as synchronized when there are no pending commits or changesets', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '1.2.0',
          baselineCommit: 'a1b2c3d4e5f6789',
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '0.8.1',
          baselineCommit: 'd4e5f6g7h8i9012',
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should warn when there are commits without a corresponding changeset', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'warning',
          message: '',
          commitCount: 3,
          packageVersion: '1.2.0',
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: add new component system' },
            { sha: '5d6e7f8g9h0i123', message: 'fix: resolve navigation bug' },
            { sha: '9g0h1i2j3k4l567', message: 'docs: update README examples' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '0.8.1',
          baselineCommit: 'd4e5f6g7h8i9012',
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should display a pending changeset when commits are present and a changeset is detected', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'pending',
          message: '',
          commitCount: 2,
          packageVersion: '1.2.0',
          changeset: {
            fileName: 'fantastic-dogs-dance.md',
            bumpType: 'minor',
          },
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: implement new authentication flow' },
            { sha: '5d6e7f8g9h0i123', message: 'fix: handle edge cases in validation' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'warning',
          message: '',
          commitCount: 2,
          packageVersion: '0.8.1',
          commits: [
            { sha: '3j4k5l6m7n8o901', message: 'feat: improve CLI performance' },
            { sha: '7m8n9o0p1q2r345', message: 'fix: handle edge case in parsing' },
          ],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should display error messages when check errors occur for one or more packages', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'error',
          message: 'Unable to find baseline commit in git history',
          commitCount: 0,
          packageVersion: '1.2.0',
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'error',
          message: 'Git repository not found or corrupted',
          commitCount: 0,
          packageVersion: '0.8.1',
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should handle a complex mixed situation with pending changesets and warnings', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'pending',
          message: '',
          commitCount: 1,
          packageVersion: '1.2.0',
          changeset: {
            fileName: 'amazing-cats-jump.md',
            bumpType: 'patch',
          },
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'fix: critical security vulnerability patch' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'warning',
          message: '',
          commitCount: 8,
          packageVersion: '0.8.1',
          commits: [
            { sha: '1p2q3r4s5t6u789', message: 'feat: add new validation system' },
            { sha: '5s6t7u8v9w0x123', message: 'fix: memory leak in file watcher' },
            { sha: '9v0w1x2y3z4a567', message: 'refactor: simplify config handling' },
            { sha: '3y4z5a6b7c8d901', message: 'test: add comprehensive integration tests' },
            { sha: '7b8c9d0e1f2g345', message: 'docs: update API documentation' },
            { sha: '1e2f3g4h5i6j789', message: 'feat: implement plugin architecture' },
            { sha: '5h6i7j8k9l0m123', message: 'perf: optimize bundling process' },
            { sha: '9k0l1m2n3o4p567', message: 'chore: update all dependencies' },
          ],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should display a warning when there are many commits without a changeset (limit case)', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'warning',
          message: '',
          commitCount: 12,
          packageVersion: '1.2.0',
          commits: [
            { sha: '1e2f3g4h5i6j789', message: 'feat: major refactoring of core system' },
            { sha: '5h6i7j8k9l0m123', message: 'feat: add advanced routing capabilities' },
            { sha: '9k0l1m2n3o4p567', message: 'fix: critical security vulnerability' },
            { sha: '3n4o5p6q7r8s901', message: 'fix: performance issues in large datasets' },
            { sha: '7q8r9s0t1u2v345', message: 'feat: implement new authentication flow' },
            { sha: '1t2u3v4w5x6y789', message: 'refactor: modernize codebase structure' },
            { sha: '5w6x7y8z9a0b123', message: 'test: comprehensive test coverage' },
            { sha: '9z0a1b2c3d4e567', message: 'docs: complete API documentation' },
            { sha: '3c4d5e6f7g8h901', message: 'fix: edge cases in error handling' },
            { sha: '7f8g9h0i1j2k345', message: 'feat: add internationalization support' },
            { sha: '1i2j3k4l5m6n789', message: 'chore: update dependencies' },
            { sha: '5l6m7n8o9p0q123', message: 'style: code formatting improvements' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '0.8.1',
          baselineCommit: 'd4e5f6g7h8i9012',
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should show a pending changeset when only part of the commits are covered (partial changeset)', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'pending',
          message: '',
          commitCount: 4,
          packageVersion: '1.2.0',
          changeset: {
            fileName: 'strong-garlics-grab.md',
            bumpType: 'minor',
          },
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: add user management system' },
            { sha: '5d6e7f8g9h0i123', message: 'fix: resolve authentication bug' },
            { sha: '9g0h1i2j3k4l567', message: 'chore: update test configuration' },
            { sha: '3k4l5m6n7o8p901', message: 'docs: update deployment guide' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'pending',
          message: '',
          commitCount: 1,
          packageVersion: '0.8.1',
          changeset: {
            fileName: 'shy-things-stare.md',
            bumpType: 'patch',
          },
          commits: [
            { sha: '7m8n9o0p1q2r345', message: 'fix: handle CLI argument parsing edge case' },
          ],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should warn when a package has commits but no version (no packageVersion field)', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'warning',
          message: '',
          commitCount: 2,
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: initial project setup' },
            { sha: '5d6e7f8g9h0i123', message: 'docs: add initial README' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should group packages by the same changeset file for multi-package changesets', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'pending',
          message: '',
          commitCount: 2,
          packageVersion: '1.0.0',
          changeset: {
            fileName: 'auth-system.md',
            bumpType: 'minor',
          },
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: implement authentication system' },
            { sha: '5d6e7f8g9h0i123', message: 'feat: add user permissions' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'pending',
          message: '',
          commitCount: 1,
          packageVersion: '0.8.1',
          changeset: {
            fileName: 'auth-system.md',
            bumpType: 'patch',
          },
          commits: [{ sha: '3j4k5l6m7n8o901', message: 'fix: CLI auth integration' }],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should handle multiple multi-package changesets and group them correctly', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'pending',
          message: '',
          commitCount: 2,
          packageVersion: '1.0.0',
          changeset: {
            fileName: 'auth-system.md',
            bumpType: 'minor',
          },
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: implement authentication system' },
            { sha: '5d6e7f8g9h0i123', message: 'feat: add user permissions' },
          ],
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'pending',
          message: '',
          commitCount: 3,
          packageVersion: '0.8.1',
          changeset: {
            fileName: 'shared-helpers.md',
            bumpType: 'patch',
          },
          commits: [
            { sha: '3j4k5l6m7n8o901', message: 'fix: CLI helper functions' },
            { sha: '7m8n9o0p1q2r345', message: 'refactor: shared utilities' },
            { sha: '1b2c3d4e5f6g789', message: 'test: add helper tests' },
          ],
        } as CheckResult,
      },
      {
        label: 'Utils',
        result: {
          status: 'pending',
          message: '',
          commitCount: 1,
          packageVersion: '0.3.2',
          changeset: {
            fileName: 'auth-system.md',
            bumpType: 'patch',
          },
          commits: [{ sha: '9x0y1z2a3b4c567', message: 'fix: auth utility functions' }],
        } as CheckResult,
      },
      {
        label: 'Types',
        result: {
          status: 'pending',
          message: '',
          commitCount: 1,
          packageVersion: '1.1.0',
          changeset: {
            fileName: 'shared-helpers.md',
            bumpType: 'minor',
          },
          commits: [{ sha: '5e6f7g8h9i0j123', message: 'feat: add helper types' }],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should prioritize ERROR status over other statuses when present', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'error',
          message: 'Git repository not found',
          commitCount: 0,
          packageVersion: '1.0.0',
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'pending',
          message: '',
          commitCount: 2,
          packageVersion: '0.8.1',
          changeset: {
            fileName: 'awesome-fix.md',
            bumpType: 'patch',
          },
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: new feature' },
            { sha: '5d6e7f8g9h0i123', message: 'fix: bug fix' },
          ],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should prioritize WARNING status over SYNC when no errors are present', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '1.0.0',
          baselineCommit: 'a1b2c3d4e5f6789',
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'warning',
          message: '',
          commitCount: 3,
          packageVersion: '0.8.1',
          commits: [
            { sha: '1a2b3c4d5e6f789', message: 'feat: needs changeset' },
            { sha: '5d6e7f8g9h0i123', message: 'fix: needs changeset' },
            { sha: '9g0h1i2j3k4l567', message: 'docs: needs changeset' },
          ],
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });

  test('should show all packages as synchronized when all statuses are SYNC', async () => {
    const data = [
      {
        label: 'Starter',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '1.0.0',
          baselineCommit: 'a1b2c3d4e5f6789',
        } as CheckResult,
      },
      {
        label: 'CLI',
        result: {
          status: 'sync',
          message: '',
          commitCount: 0,
          packageVersion: '0.8.1',
          baselineCommit: 'd4e5f6g7h8i9012',
        } as CheckResult,
      },
    ];
    expect(normalizeReport(await formatSyncOutput(data, '/fake/workspace/root'))).toMatchSnapshot();
  });
});

describe('formatBaselineLog', () => {
  test('should format basic baseline log for non-Starter config', async () => {
    const config: SyncCheckConfig = {
      label: 'CLI',
      baseline: () => Promise.resolve('abc123456789'),
      commitPath: 'packages/cli/',
      messages: {
        notFound: '',
        inSync: '',
        unreleased: '',
        errorPrefix: '',
      },
    };

    const result = await formatBaselineLog(config, 'abc123456789', '/test/workspace');
    expect(result).toBe('📦 [CLI] Checking against baseline (commit/tag abc1234)');
  });

  test('should format basic baseline log for any package config', async () => {
    const config: SyncCheckConfig = {
      label: 'Starter',
      baseline: () => Promise.resolve('abc123456789'),
      commitPath: 'apps/starter/',
      messages: {
        notFound: '',
        inSync: '',
        unreleased: '',
        errorPrefix: '',
      },
    };

    const result = await formatBaselineLog(config, 'abc123456789', '/test/workspace');
    expect(result).toBe('📦 [Starter] Checking against baseline (commit/tag abc1234)');
  });
});
