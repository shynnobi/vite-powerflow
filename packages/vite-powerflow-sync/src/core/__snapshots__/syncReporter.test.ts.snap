// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatSyncOutput > should display a pending changeset when commits are present and a changeset is detected 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0)",
  "   📄 Changeset: fantastic-dogs-dance.md (minor)",
  "   📊 Coverage: 0/2 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [CLI] (v0.8.1)",
  "   ❌ No changeset found",
  "   📋 2 unreleased commits need changeset:",
  "      • 3j4k5l6 feat: improve CLI performance",
  "      • 7m8n9o0 fix: handle edge case in parsing",
  "   ⚠️ Require changeset (2 unreleased commits)",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟠 WARNING",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should display a warning when there are many commits without a changeset (limit case) 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0)",
  "   ❌ No changeset found",
  "   📋 12 unreleased commits need changeset:",
  "      • 1e2f3g4 feat: major refactoring of core system",
  "      • 5h6i7j8 feat: add advanced routing capabilities",
  "      • 9k0l1m2 fix: critical security vulnerability",
  "      • 3n4o5p6 fix: performance issues in large datasets",
  "      • 7q8r9s0 feat: implement new authentication flow",
  "      • ... and 7 more",
  "   ⚠️ Require changeset (12 unreleased commits)",
  "",
  "📦 [CLI] (v0.8.1) - baseline d4e5f6g",
  "   ✅ Package in sync",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟠 WARNING",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should display error messages when check errors occur for one or more packages 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0)",
  "   ❌ Check failed - Unable to find baseline commit in git history",
  "",
  "📦 [CLI] (v0.8.1)",
  "   ❌ Check failed - Git repository not found or corrupted",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🔴 ERROR",
  "📋 Summary: All 2 packages need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should group packages by the same changeset file for multi-package changesets 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.0.0)",
  "   📄 Changeset: auth-system.md (minor)",
  "   📊 Coverage: 0/2 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [CLI] (v0.8.1)",
  "   📄 Changeset: auth-system.md (patch)",
  "   📊 Coverage: 0/1 commits covered",
  "   🎯 Ready for release",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟡 PENDING",
  "📋 Summary: All 2 packages ready",
  "🔗 auth-system.md: Starter (minor), CLI (patch)",
  "",
  "",
]
`;

exports[`formatSyncOutput > should handle a complex mixed situation with pending changesets and warnings 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0)",
  "   📄 Changeset: amazing-cats-jump.md (patch)",
  "   📊 Coverage: 0/1 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [CLI] (v0.8.1)",
  "   ❌ No changeset found",
  "   📋 8 unreleased commits need changeset:",
  "      • 1p2q3r4 feat: add new validation system",
  "      • 5s6t7u8 fix: memory leak in file watcher",
  "      • 9v0w1x2 refactor: simplify config handling",
  "      • 3y4z5a6 test: add comprehensive integration tests",
  "      • 7b8c9d0 docs: update API documentation",
  "      • ... and 3 more",
  "   ⚠️ Require changeset (8 unreleased commits)",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟠 WARNING",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should handle multiple multi-package changesets and group them correctly 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.0.0)",
  "   📄 Changeset: auth-system.md (minor)",
  "   📊 Coverage: 0/2 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [CLI] (v0.8.1)",
  "   📄 Changeset: shared-helpers.md (patch)",
  "   📊 Coverage: 0/3 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [Utils] (v0.3.2)",
  "   📄 Changeset: auth-system.md (patch)",
  "   📊 Coverage: 0/1 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [Types] (v1.1.0)",
  "   📄 Changeset: shared-helpers.md (minor)",
  "   📊 Coverage: 0/1 commits covered",
  "   🎯 Ready for release",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟡 PENDING",
  "📋 Summary: All 4 packages ready",
  "🔗 auth-system.md: Starter (minor), Utils (patch)",
  "🔗 shared-helpers.md: CLI (patch), Types (minor)",
  "",
  "",
]
`;

exports[`formatSyncOutput > should prioritize ERROR status over other statuses when present 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.0.0)",
  "   ❌ Check failed - Git repository not found",
  "",
  "📦 [CLI] (v0.8.1)",
  "   📄 Changeset: awesome-fix.md (patch)",
  "   📊 Coverage: 0/2 commits covered",
  "   🎯 Ready for release",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🔴 ERROR",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should prioritize WARNING status over SYNC when no errors are present 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.0.0) - baseline a1b2c3d",
  "   ✅ Package in sync",
  "",
  "📦 [CLI] (v0.8.1)",
  "   ❌ No changeset found",
  "   📋 3 unreleased commits need changeset:",
  "      • 1a2b3c4 feat: needs changeset",
  "      • 5d6e7f8 fix: needs changeset",
  "      • 9g0h1i2 docs: needs changeset",
  "   ⚠️ Require changeset (3 unreleased commits)",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟠 WARNING",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should show a pending changeset when only part of the commits are covered (partial changeset) 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0)",
  "   📄 Changeset: strong-garlics-grab.md (minor)",
  "   📊 Coverage: 0/4 commits covered",
  "   🎯 Ready for release",
  "",
  "📦 [CLI] (v0.8.1)",
  "   📄 Changeset: shy-things-stare.md (patch)",
  "   📊 Coverage: 0/1 commits covered",
  "   🎯 Ready for release",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟡 PENDING",
  "📋 Summary: All 2 packages ready",
  "",
  "",
]
`;

exports[`formatSyncOutput > should show all packages as synchronized when all statuses are SYNC 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.0.0) - baseline a1b2c3d",
  "   ✅ Package in sync",
  "",
  "📦 [CLI] (v0.8.1) - baseline d4e5f6g",
  "   ✅ Package in sync",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟢 SYNC",
  "📋 Summary: All 2 packages ready",
  "",
  "",
]
`;

exports[`formatSyncOutput > should show all packages as synchronized when there are no pending commits or changesets 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0) - baseline a1b2c3d",
  "   ✅ Package in sync",
  "",
  "📦 [CLI] (v0.8.1) - baseline d4e5f6g",
  "   ✅ Package in sync",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟢 SYNC",
  "📋 Summary: All 2 packages ready",
  "",
  "",
]
`;

exports[`formatSyncOutput > should warn when a package has commits but no version (no packageVersion field) 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter]",
  "   ❌ No changeset found",
  "   📋 2 unreleased commits need changeset:",
  "      • 1a2b3c4 feat: initial project setup",
  "      • 5d6e7f8 docs: add initial README",
  "   ⚠️ Require changeset (2 unreleased commits)",
  "",
  "📦 [CLI]",
  "   ✅ Package in sync",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟠 WARNING",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;

exports[`formatSyncOutput > should warn when there are commits without a corresponding changeset 1`] = `
[
  "🔄 Sync Status Report - [TIME]",
  "═══════════════════════════════════════════════════════════",
  "",
  "📦 [Starter] (v1.2.0)",
  "   ❌ No changeset found",
  "   📋 3 unreleased commits need changeset:",
  "      • 1a2b3c4 feat: add new component system",
  "      • 5d6e7f8 fix: resolve navigation bug",
  "      • 9g0h1i2 docs: update README examples",
  "   ⚠️ Require changeset (3 unreleased commits)",
  "",
  "📦 [CLI] (v0.8.1) - baseline d4e5f6g",
  "   ✅ Package in sync",
  "",
  "═══════════════════════════════════════════════════════════",
  "🔄 Status: 🟠 WARNING",
  "📋 Summary: 1 package ready, 1 package need attention",
  "",
  "",
]
`;
